File: Template Method/Golang/TwoFactorPhone.go
package main

import "fmt"

type TwoFactorPhone struct {}

func (twoFactor *TwoFactorPhone) getData(username string)  {
	fmt.Println("Getting phone data for", username)
}

func (twoFactor *TwoFactorPhone) sendNotification(username string)  {
	fmt.Println("Sending sms to", username)
}

func (twoFactor *TwoFactorPhone) match(username string)  {
	fmt.Println("Matching sms code from", username)
}

File: Template Method/Javascript/TwoFactorPhone.js
import TwoFactorAuthentication from "./TwoFactorAuthentication.js";

class TwoFactorPhone extends TwoFactorAuthentication {
  getData(username) {
    console.log(`Getting phone number data for ${username}`);
  }

  sendNotification(username) {
    console.log(`Sending SMS notification to ${username}`);
  }

  match(username) {
    console.log(`Matching ${username} SMS code`);
  }
}

export default TwoFactorPhone;
